### PR TITLE
Support optional ALLOW_NO_DEVICE=true in upsd

### PR DESCRIPTION
**Pull request by jimklimov**
Allow curious admins to start their nut server deliberately without any device configuration sections yet defined. A subsequent `reload` can have this daemon learn about changed configurations, but from the get-go first boot it can serve the NUT protocol with whatever it knows (supposedly, a clean "nothing here!" statement).

Also includes a fix for reload from https://github.com/42ity/nut/pull/64

UPDATE: After some evolution of this PR and its counterpart in 42ity, the setting was renamed to a more agreeable spelling, and it is possible to set it with `upsd.conf` (probably the recommended manner now) or with an environment variable passed to `upsd`, at higher priority than hardcoded or configured value (allowing to tweak it for custom runs, dev-tests, etc. - and for the originally proposed setting via `nut.conf` file which is sourced by systemd and SMF units and most init-scripts).

### CODE REVIEW

**Comment by jimklimov** [comment]
In testing, behaves as expected:
* the `upsd` server (wrapped into `nut-server.service` unit with `/etc/nut/nut.conf` having the uncommented variable) starts and logs the warning about zero device sections
* the `nut-driver-enumerator` is able to trigger reloads of `nut-server`, these are logged (discovering again that there are zero devices) and the same daemon process is running
* adding a device section to `/etc/nut/ups.conf` caused the config reload (via nut-driver-enumerator) and connection to the newly started driver (with service unit just defined by NDE)
** `upsc -l` shows the new `test` device,
** and `upsc test` shows its data
* commenting away the section causes the reload (via NDE) and warning about zero devices
** `upsc -l` shows no output, zero exit code
** `upsc test` says `Error: Unknown UPS` with exit code ` for error
* the same `upsd` PID as was originally started, keeps running in the `nut-server.service`

**Comment by jimklimov** [comment]
For a holistic solution, updated with https://github.com/42ity/nut/pull/109 to support also setting this configuration value "properly" via upsd.conf. The code still supports an envvar as a higher-priority option, this can be used in production (e.g. via nut.conf and init-scripts or service units that honor it) or in development to run your built `upsd` in this special manner. Might be used in unit-tests for example...

**Comment by jimklimov** [comment]
Bump? @clepple  @aquette  @zykh : I intend to merge lingering PRs the next time I hover over them, maybe in a week or so, so here's your last chance to approve or bike-shed :)

File: scripts/systemd/nut-driver-enumerator.service.in [hunk 1]
```diff
@@ -21,6 +21,7 @@ Type=oneshot
 # nut-server which is the primary concerned dependency at the moment, so we
 # don't want it to fail the unit (when it can't restart).
 Environment=REPORT_RESTART_42=no
+EnvironmentFile=-@CONFPATH@/nut.conf
```
Comment (clepple) [inline scripts/systemd/nut-driver-enumerator.service.in:24]: Will the `... ; export ALLOW_NO_DEVICE` shell syntax work with systemd's `EnvironmentFile` parser? Have not tested myself.

https://www.freedesktop.org/software/systemd/man/systemd.exec.html

File: docs/man/upsd.conf.txt [hunk 2]
```diff
@@ -32,6 +32,19 @@ When instant commands and variables setting status tracking is enabled, status
 execution information are kept during this amount of time, and then cleaned up.
 This defaults to 3600 (1 hour).
 
+"ALLOW_NO_DEVICE 'Boolean'"::
```
Comment (clepple) [inline docs/man/upsd.conf.txt:35]: I would mention the environment variable as well, and the fact that it takes precedence over the conf file (realizing that this is mentioned in the sample file)

**Review by clepple — COMMENTED** [review]
(see inline)

File: scripts/systemd/nut-driver-enumerator.service.in (hunk 1 above)
Comment (jimklimov) [reply scripts/systemd/nut-driver-enumerator.service.in:24]: Very good catch, thanks. By coincidence, we check the first 4/5 chars of the envvar so find `true`/`false` anyway, so this slipped past the tests. But yes, systemd puts everything after `=` into the envvar value:

````
root@debian8:~# cat /run/systemd/system/envtest.service 
[Service]
EnvironmentFile=/run/envfile
ExecStart=/bin/sh -c 'set | grep TEST'
Type=oneshot
````

````
root@debian8:~# cat /run/envfile
TEST=123; export TEST
TESTML=456
export TESTML
````

Result of experiment: TEST is weird and wrong, TESTML is as expected:
````
# systemctl status envtest
Oct 12 11:07:59 debian8 systemd[1]: Starting envtest.service...
Oct 12 11:07:59 debian8 sh[26737]: TEST='123; export TEST'
Oct 12 11:07:59 debian8 systemd[1]: Started envtest.service.
Oct 12 11:08:40 debian8 systemd[1]: Starting envtest.service...
Oct 12 11:08:40 debian8 sh[26771]: TEST='123; export TEST'
Oct 12 11:08:40 debian8 sh[26771]: TESTML='456'
Oct 12 11:08:40 debian8 systemd[1]: Started envtest.service.
````

File: docs/man/upsd.conf.txt (hunk 2 above)
Comment (jimklimov) [reply docs/man/upsd.conf.txt:35]: Thanks, done.

File: scripts/systemd/nut-driver-enumerator.service.in (hunk 1 above)
Comment (jimklimov) [reply scripts/systemd/nut-driver-enumerator.service.in:24]: Split the example into two lines, and added a general comment about expected syntax for the file.

**Comment by aquette** [comment]
a side comment, vaguely related, with Debian and systemd, to shed a bit more light on the topic:
https://lintian.debian.org/tags/init.d-script-should-always-start-service.html

in other words, either NUT is not configured, and all systemd units are disabled (or better masked) ; otherwise NUT is configured (fully or just with ALLOW_NO_DEVICE) and at least upsd unit is startable.  Otheriwse, systemd units will fail to start if ALLOW_NO_DEVICE is not active and set to true by default.

**Comment by jimklimov** [comment]
@aquette : would it make sense to suicide? :) Amend all our units to read `EnvironmentFile=-@CONFPATH@/nut.conf` and `ExecStartPre=/bin/sh 'if test "$MODE" = none ; then /bin/systemctl disable %n ; /bin/systemctl mask %n ; /bin/systemctl --no-block stop %n ; fi'` or something of the sort?

**Comment by jimklimov** [comment]
I guess I'll leave the alignment with systemd recommendation as a separate issue that anyone interested can tackle sometime...

**Pull request merged**